crates/compiler/builtins/roc/Dict.roc

```diff
@@ -339,25 +484,342 @@ keepShared = \@Dict xs, ys ->
 ##
 ##     expect Dict.removeAll first second == expected
-removeAll : Dict k v, Dict k v -> Dict k v | k has Eq
-removeAll = \xs, @Dict ys ->
-    List.walk ys xs (\state, Pair k _ -> Dict.remove state k)
-
-## Internal helper function to insert a new association
-##
-## Precondition: `k` should not exist in the Dict yet.
-insertFresh : Dict k v, k, v -> Dict k v
-insertFresh = \@Dict list, k, v ->
-    list
-    |> List.append (Pair k v)
-    |> @Dict
-
-insertIfVacant : Dict k v, k, v -> Dict k v | k has Eq
-insertIfVacant = \dict, key, value ->
-    if Dict.contains dict key then
-        dict
+removeAll : Dict k v, Dict k v -> Dict k v | k has Hash & Eq
+removeAll = \xs, ys ->
+    walk ys xs (\state, k, _ -> remove state k)
+
+swapAndUpdateDataIndex : Dict k v, Nat, Nat -> Dict k v | k has Hash & Eq
+swapAndUpdateDataIndex = \@Dict { metadata, dataIndices, data, size }, removedIndex, lastIndex ->
+    (T key _) = listGetUnsafe data lastIndex
+    hashKey =
+        createLowLevelHasher {}
+        |> Hash.hash key
+        |> complete
+    h1Key = h1 hashKey
+    h2Key = h2 hashKey
+    probe = newProbe h1Key (div8 (List.len metadata))
+
+    when findIndexHelper metadata dataIndices data h2Key key probe 0 is
+        Ok index ->
+            dataIndex = listGetUnsafe dataIndices removedIndex
+            # Swap and remove data.
+            nextData =
+                data
+                |> List.swap dataIndex lastIndex
+                |> List.dropLast
+
+            @Dict {
+                # Set old metadata as deleted.
+                metadata: List.set metadata removedIndex deletedSlot,
+                # Update index of swaped element.
+                dataIndices: List.set dataIndices index dataIndex,
+                data: nextData,
+                size: size - 1,
+            }
+
+        Err NotFound ->
+            # This should be impossible.
+            crash "unreachable state in dict swapAndUpdateDataIndex hit. Definitely a standard library bug."
+
+insertNotFoundHelper : Dict k v, k, v, U64, I8 -> Dict k v
+insertNotFoundHelper = \@Dict { metadata, dataIndices, data, size }, key, value, h1Key, h2Key ->
+    probe = newProbe h1Key (div8 (List.len metadata))
+    index = nextEmptyOrDeletedHelper metadata probe 0
+    dataIndex = List.len data
+    nextData = List.append data (T key value)
+
+    @Dict {
+        metadata: List.set metadata index h2Key,
+        dataIndices: List.set dataIndices index dataIndex,
+        data: nextData,
+        size,
+    }
+
+nextEmptyOrDeletedHelper : List I8, Probe, Nat -> Nat
+nextEmptyOrDeletedHelper = \metadata, probe, offset ->
+    # For inserting, we can use deleted indices.
+    index = Num.addWrap (mul8 probe.slotIndex) offset
+
+    md = listGetUnsafe metadata index
+
+    if md < 0 then
+        # Empty or deleted slot, no possibility of the element.
+        index
+    else if offset == 7 then
+        nextEmptyOrDeletedHelper metadata (nextProbe probe) 0
```

Comment:
Can save this for later as with the other todos in this file - can docs for what the significance of `7` here and elsewhere is be added?

Reply:
Ah, yeah. Probably should have made that a constant for clarity. It would have made a lot more sense if I had gotten groups working correctly and fast. Essentially the group size is 8. So that is group size - 1. With the grouping pr, we should just be using a U64 directly, so it will have 8 U8s in it. Traversing that will clarify the count.